File: SummerCal/Services/SmartNotificationService.swift
import Foundation
import SwiftData

final class SmartNotificationService {
    private let notificationService: NotificationService
    private let freeDayService: FreeDayDetectionService

    init(notificationService: NotificationService = .shared,
         freeDayService: FreeDayDetectionService = FreeDayDetectionService()) {
        self.notificationService = notificationService
        self.freeDayService = freeDayService
    }

    func runDailyPipeline(
        for date: Date,
        modelContext: ModelContext,
        weatherService: WeatherService? = nil,
        placesService: PlacesService? = nil
    ) async {
        let settings = UserSettings.current(in: modelContext)

        let eventDescriptor = FetchDescriptor<CalendarEvent>(sortBy: [SortDescriptor(\.startDate)])
        guard let allEvents = try? modelContext.fetch(eventDescriptor) else { return }

        let dayEvents = allEvents.filter { Calendar.current.isDate($0.startDate, inSameDayAs: date) }

        let ruleDescriptor = FetchDescriptor<SmartNotificationRule>()
        guard let rules = try? modelContext.fetch(ruleDescriptor) else { return }

        let todayLogs = (try? modelContext.fetch(FetchDescriptor<NotificationLog>()))?.filter {
            Calendar.current.isDate($0.scheduledFor, inSameDayAs: date)
        } ?? []

        for rule in rules where rule.isEnabled {
            if todayLogs.filter({ $0.kind == rule.kind }).count >= rule.maxPerDay { continue }
            if isInQuietHours(settings: settings) { continue }

            switch rule.kind {
            case .upcomingEvent:
                await scheduleUpcomingEventNotifications(events: dayEvents, notes: nil)
            case .freeDay:
                let freeWindows = freeDayService.freeWindows(for: date, events: dayEvents)
                if freeDayService.shouldNotifyFreeDay(freeWindows: freeWindows, thresholdHours: Double(settings.freeDayThresholdHours)) {
                    let title = "Free Day"
                    let body = freeDayService.notificationBody(for: freeWindows, weather: nil)
                    await scheduleAndLog(kind: .freeDay, title: title, body: body, date: date, rule: rule, context: modelContext)
                }
            case .freeAfternoon:
                let freeWindows = freeDayService.freeWindows(for: date, events: dayEvents)
                let afternoonFree = freeWindows.filter { win in
                    let hour = Calendar.current.component(.hour, from: win.start)
                    return hour >= 12
                }
                if freeDayService.shouldNotifyFreeDay(freeWindows: afternoonFree, thresholdHours: 2) {
                    await scheduleAndLog(
                        kind: .freeAfternoon,
                        title: "Free Afternoon",
                        body: "You have free time this afternoon. Good time for errands, gym, or a cafe work block.",
                        date: date,
                        rule: rule,
                        context: modelContext
                    )
                }
            case .weatherSummary:
                let title = "Weather Today"
                let todayWeather = fetchTodaysWeather(context: modelContext)
                let body = weatherService?.generateWeatherSummary(weather: todayWeather) ?? generateLocalWeatherSummary(from: todayWeather)
                await scheduleAndLog(kind: .weatherSummary, title: title, body: body, date: date, rule: rule, context: modelContext)
            case .weatherAlert:
                let todayWeather = fetchTodaysWeather(context: modelContext)
                await scheduleWeatherAlertsIfNeeded(weather: todayWeather, settings: settings, date: date, rule: rule, context: modelContext)
            case .eventPrep:
                break
            case .moneyReminder:
                if (settings.monthlyIncomeGoal ?? 0) > 0 {
                    await scheduleAndLog(
                        kind: .moneyReminder,
                        title: "Money Check",
                        body: "How's your monthly goal going?",
                        date: date,
                        rule: rule,
                        context: modelContext
                    )
                }
            }
        }
    }

    // MARK: - Weather helpers

    private func fetchTodaysWeather(context: ModelContext) -> WeatherSnapshot? {
        let now = Date()
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: now)
        let endOfDay = calendar.date(byAdding: .day, value: 1, to: startOfDay)!

        let descriptor = FetchDescriptor<WeatherSnapshot>(
            predicate: #Predicate { snapshot in
                snapshot.forecastDate >= startOfDay && snapshot.forecastDate < endOfDay
            },
            sortBy: [SortDescriptor(\.fetchedAt, order: .reverse)]
        )
        var results = (try? context.fetch(descriptor)) ?? []
        if results.isEmpty {
            let allDescriptor = FetchDescriptor<WeatherSnapshot>(
                sortBy: [SortDescriptor(\.fetchedAt, order: .reverse)]
            )
            results = (try? context.fetch(allDescriptor)) ?? []
        }
        return results.first
    }

    private func generateLocalWeatherSummary(from snapshot: WeatherSnapshot?) -> String {
        guard let w = snapshot else { return "Check the weather tab for today's forecast." }
        var parts = ["\(w.condition), \(String(format: "%.0f", w.temperatureCelsius))°C"]
        if let feels = w.feelsLikeCelsius {
            parts.append("feels like \(String(format: "%.0f", feels))°C")
        }
        if w.precipitationChance > 0.3 {
            parts.append("\(Int(w.precipitationChance * 100))% rain")
        }
        if let wind = w.windSpeedKph, wind > 15 {
            parts.append("wind \(String(format: "%.0f", wind)) km/h")
        }
        if let uv = w.uvIndex, uv >= 6 {
            parts.append("UV Index \(uv)")
        }
        return parts.joined(separator: ", ")
    }

    private func scheduleWeatherAlertsIfNeeded(
        weather: WeatherSnapshot?,
        settings: UserSettings,
        date: Date,
        rule: SmartNotificationRule,
        context: ModelContext
    ) async {
        guard let w = weather else { return }
        var alerts: [String] = []

        if settings.weatherAlertsEnabled {
            if w.precipitationChance >= settings.rainThreshold {
                alerts.append("Rain alert: \(Int(w.precipitationChance * 100))% chance today")
            }
            if w.temperatureCelsius >= settings.heatThresholdCelsius {
                alerts.append("Heat alert: \(String(format: "%.0f", w.temperatureCelsius))°C exceeds \(Int(settings.heatThresholdCelsius))°C")
            }
            if w.temperatureCelsius <= settings.coldThresholdCelsius {
                alerts.append("Cold alert: \(String(format: "%.0f", w.temperatureCelsius))°C is below \(Int(settings.coldThresholdCelsius))°C")
            }
        }

        for alert in alerts {
            await scheduleAndLog(
                kind: .weatherAlert,
                title: "Weather Alert",
                body: alert,
                date: date,
                rule: rule,
                context: context
            )
        }
    }

    // MARK: - Event notifications

    private func scheduleUpcomingEventNotifications(events: [CalendarEvent], notes: [EventNote]?) async {
        for event in events {
            guard event.notificationEnabled else { continue }
            _ = await notificationService.scheduleEventReminder(
                event: event,
                minutesBefore: event.reminderMinutesBefore,
                notes: notes?.filter { $0.eventId == event.id }
            )
        }
    }

    private func scheduleAndLog(
        kind: SmartNotificationKind,
        title: String,
        body: String,
        date: Date,
        rule: SmartNotificationRule,
        context: ModelContext
    ) async {
        let notificationId = "smart_\(kind.rawValue)_\(UUID().uuidString)"
        let scheduleDate = nextScheduledDate(for: rule, on: date)

        let scheduled = await notificationService.scheduleSmartNotification(
            id: notificationId,
            title: title,
            body: body,
            date: scheduleDate,
            categoryIdentifier: kind == .freeDay ? "FREE_DAY" : nil
        )
        guard scheduled else { return }

        let log = NotificationLog(
            notificationId: notificationId,
            kind: kind,
            title: title,
            body: body,
            scheduledFor: scheduleDate
        )
        context.insert(log)
        try? context.save()
    }

    private func nextScheduledDate(for rule: SmartNotificationRule, on date: Date) -> Date {
        let calendar = Calendar.current
        let preferred = calendar.date(
            bySettingHour: rule.preferredHour,
            minute: rule.preferredMinute,
            second: 0,
            of: date
        ) ?? date
        let minimumFutureDate = Date().addingTimeInterval(60)

        if preferred > minimumFutureDate {
            return preferred
        }

        if calendar.isDate(date, inSameDayAs: Date()) {
            return minimumFutureDate
        }

        return calendar.date(byAdding: .day, value: 1, to: preferred) ?? minimumFutureDate
    }

    private func isInQuietHours(settings: UserSettings) -> Bool {
        let hour = Calendar.current.component(.hour, from: Date())
        let quietStart = settings.quietHoursStart
        let quietEnd = settings.quietHoursEnd
        if quietStart < quietEnd {
            return hour >= quietStart && hour < quietEnd
        } else {
            return hour >= quietStart || hour < quietEnd
        }
    }
}

File: SummerCalExpo/src/services/SmartNotificationService.ts
import { CalendarEvent, WeatherSnapshot, SmartNotificationRule, SmartNotificationKind, ActivitySuggestion, NotificationLog, UserSettings } from '../models';
import { NotificationService } from './NotificationService';
import { FreeDayDetectionService } from './FreeDayDetectionService';
import { ActivitySuggestionService } from './ActivitySuggestionService';
import { WeatherService } from './WeatherService';
import * as SQLite from 'expo-sqlite';
import { v4 as uuidv4 } from 'uuid';

export class SmartNotificationService {
  private notificationService: NotificationService;
  private freeDayService: FreeDayDetectionService;
  private activityService: ActivitySuggestionService;
  private weatherService: WeatherService;

  constructor(
    notificationService?: NotificationService,
    freeDayService?: FreeDayDetectionService,
    activityService?: ActivitySuggestionService,
    weatherService?: WeatherService
  ) {
    this.notificationService = notificationService ?? new NotificationService();
    this.freeDayService = freeDayService ?? new FreeDayDetectionService();
    this.activityService = activityService ?? new ActivitySuggestionService();
    this.weatherService = weatherService ?? new WeatherService();
  }

  async runDailyPipeline(
    date: Date,
    db: SQLite.SQLiteDatabase,
    events: CalendarEvent[],
    settings: UserSettings,
    rules: SmartNotificationRule[],
    weather?: WeatherSnapshot
  ): Promise<NotificationLog[]> {
    const logs: NotificationLog[] = [];
    const now = new Date().toISOString();
    const nowHour = date.getHours();

    const todaySent = await this.todaySentCount(db, date);
    const maxTotal = settings.maxSmartNotificationsPerDay;

    if (todaySent >= maxTotal) {
      this.addQuietLog(date, 'global_limit', logs, now);
      return logs;
    }

    let remaining = maxTotal - todaySent;

    for (const rule of rules) {
      if (remaining <= 0) break;
      if (!rule.isEnabled) continue;
      if (rule.preferredHour !== 0 && nowHour < rule.preferredHour) continue;

      const inQuiet = this.isQuietHours(nowHour, rule.quietHoursStart, rule.quietHoursEnd);
      if (inQuiet) continue;

      const ruleCount = await this.todayRuleCount(db, rule.kind, date);
      if (ruleCount >= rule.maxPerDay) continue;

      switch (rule.kind) {
        case SmartNotificationKind.freeDay: {
          const freeWindows = this.freeDayService.freeWindows(date, events);
          const freeHours = freeWindows.reduce((s, w) => s + (w.end.getTime() - w.start.getTime()) / 3600000, 0);
          const threshold = settings.freeDayThresholdHours;

          if (freeHours >= threshold) {
            const notificationId = `smart_free_day_${date.toISOString().slice(0, 10)}`;
            const body = this.freeDayService.notificationBody(
              freeWindows,
              weather ? { precipitationChance: weather.precipitationChance, temperatureCelsius: weather.temperatureCelsius } : undefined
            );
            const scheduledDate = this.buildScheduleDate(date, rule.preferredHour, rule.preferredMinute);
            await this.notificationService.scheduleSmartNotification(
              notificationId, `Free afternoon — about ${Math.round(freeHours)} hours`, body, scheduledDate,
              { kind: SmartNotificationKind.freeDay }
            );
            logs.push(this.makeLog(notificationId, SmartNotificationKind.freeDay, `Free afternoon`, body, scheduledDate.toISOString(), now));
            remaining--;
          }
          break;
        }

        case SmartNotificationKind.freeAfternoon: {
          const freeWindows = this.freeDayService.freeWindows(date, events);
          const afternoonWindows = freeWindows.filter(w => w.start.getHours() >= 12 && w.start.getHours() < 17);
          const afternoonHours = afternoonWindows.reduce((s, w) => s + (w.end.getTime() - w.start.getTime()) / 3600000, 0);

          if (afternoonHours >= 2) {
            const notificationId = `smart_free_afternoon_${date.toISOString().slice(0, 10)}`;
            const body = `You have about ${Math.round(afternoonHours)} hours free this afternoon. ${weather ? `Weather: ${this.weatherService.generateWeatherSummary(weather)}` : ''}`;
            const scheduledDate = this.buildScheduleDate(date, rule.preferredHour, rule.preferredMinute);
            await this.notificationService.scheduleSmartNotification(
              notificationId, 'Free afternoon', body, scheduledDate,
              { kind: SmartNotificationKind.freeAfternoon }
            );
            logs.push(this.makeLog(notificationId, SmartNotificationKind.freeAfternoon, 'Free afternoon', body, scheduledDate.toISOString(), now));
            remaining--;
          }
          break;
        }

        case SmartNotificationKind.upcomingEvent: {
          const upcoming = events
            .filter(e => new Date(e.startDate) > date && new Date(e.startDate).getTime() - date.getTime() < 72 * 3600000)
            .sort((a, b) => new Date(a.startDate).getTime() - new Date(b.startDate).getTime());

          for (const event of upcoming.slice(0, 3)) {
            if (remaining <= 0) break;
            const notificationId = `smart_upcoming_${event.id}`;
            const timeUntil = Math.round((new Date(event.startDate).getTime() - date.getTime()) / 3600000);
            const body = `"${event.title}" ${event.location ? `at ${event.location} ` : ''}in ${timeUntil} hours`;
            const triggerDate = new Date(date);
            triggerDate.setHours(rule.preferredHour || 7, rule.preferredMinute || 0, 0, 0);
            await this.notificationService.scheduleSmartNotification(
              notificationId, 'Upcoming event', body, triggerDate,
              { kind: SmartNotificationKind.upcomingEvent, eventId: event.id }
            );
            logs.push(this.makeLog(notificationId, SmartNotificationKind.upcomingEvent, 'Upcoming event', body, triggerDate.toISOString(), now, event.id));
            remaining--;
          }
          break;
        }

        case SmartNotificationKind.weatherSummary: {
          if (!weather) break;
          const notificationId = `smart_weather_${date.toISOString().slice(0, 10)}`;
          const body = this.weatherService.generateWeatherSummary(weather);
          const scheduledDate = this.buildScheduleDate(date, rule.preferredHour, rule.preferredMinute);
          await this.notificationService.scheduleSmartNotification(
            notificationId, 'Weather today', body, scheduledDate,
            { kind: SmartNotificationKind.weatherSummary }
          );
          logs.push(this.makeLog(notificationId, SmartNotificationKind.weatherSummary, 'Weather today', body, scheduledDate.toISOString(), now));
          remaining--;
          break;
        }

        case SmartNotificationKind.weatherAlert: {
          if (!weather) break;
          let alertTitle = '';
          let alertBody = '';
          if (this.weatherService.shouldNotifyRain(weather, settings.rainThreshold)) {
            alertTitle = 'Rain alert';
            alertBody = `Rain expected today (${Math.round(weather.precipitationChance * 100)}% chance). Bring an umbrella!`;
          } else if (this.weatherService.shouldNotifyHeat(weather, settings.heatThresholdCelsius)) {
            alertTitle = 'Heat alert';
            alertBody = `High of ${Math.round(weather.temperatureCelsius)}°C today. Stay hydrated!`;
          } else if (this.weatherService.shouldNotifyCold(weather, settings.coldThresholdCelsius)) {
            alertTitle = 'Cold alert';
            alertBody = `Low of ${Math.round(weather.temperatureCelsius)}°C today. Dress warmly!`;
          }
          if (alertTitle) {
            const notificationId = `smart_weather_alert_${date.toISOString().slice(0, 10)}`;
            const scheduledDate = this.buildScheduleDate(date, rule.preferredHour || 7, rule.preferredMinute || 0);
            await this.notificationService.scheduleSmartNotification(
              notificationId, alertTitle, alertBody, scheduledDate,
              { kind: SmartNotificationKind.weatherAlert }
            );
            logs.push(this.makeLog(notificationId, SmartNotificationKind.weatherAlert, alertTitle, alertBody, scheduledDate.toISOString(), now));
            remaining--;
          }
          break;
        }

        case SmartNotificationKind.eventPrep: {
          const soonEvents = events.filter(e => {
            const diff = new Date(e.startDate).getTime() - date.getTime();
            return diff > 0 && diff < 24 * 3600000;
          });
          for (const event of soonEvents.slice(0, 2)) {
            if (remaining <= 0) break;
            const notificationId = `smart_prep_${event.id}`;
            const body = `"${event.title}" is coming up. Check notes and prepare.`;
            const scheduledDate = this.buildScheduleDate(date, rule.preferredHour || 8, rule.preferredMinute || 0);
            await this.notificationService.scheduleSmartNotification(
              notificationId, 'Event prep reminder', body, scheduledDate,
              { kind: SmartNotificationKind.eventPrep, eventId: event.id }
            );
            logs.push(this.makeLog(notificationId, SmartNotificationKind.eventPrep, 'Event prep reminder', body, scheduledDate.toISOString(), now, event.id));
            remaining--;
          }
          break;
        }

        case SmartNotificationKind.moneyReminder: {
          const notificationId = `smart_money_${date.toISOString().slice(0, 10)}`;
          const body = 'Reminder: log today\'s earnings and expenses.';
          const scheduledDate = this.buildScheduleDate(date, rule.preferredHour, rule.preferredMinute);
          await this.notificationService.scheduleSmartNotification(
            notificationId, 'Money reminder', body, scheduledDate,
            { kind: SmartNotificationKind.moneyReminder }
          );
          logs.push(this.makeLog(notificationId, SmartNotificationKind.moneyReminder, 'Money reminder', body, scheduledDate.toISOString(), now));
          remaining--;
          break;
        }
      }
    }

    return logs;
  }

  async saveLogs(logs: NotificationLog[], db: SQLite.SQLiteDatabase): Promise<void> {
    for (const log of logs) {
      await db.runAsync(
        `INSERT INTO notification_logs (id, notification_id, kind, title, body, scheduled_for, delivered_estimate, related_event_id, related_suggestion_id, created_at)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
        log.id, log.notificationId, log.kind, log.title, log.body,
        log.scheduledFor, log.deliveredEstimate ?? null,
        log.relatedEventId ?? null, log.relatedSuggestionId ?? null,
        log.createdAt
      );
    }
  }

  private async todaySentCount(db: SQLite.SQLiteDatabase, date: Date): Promise<number> {
    const day = date.toISOString().slice(0, 10);
    const row = await db.getFirstAsync<{ cnt: number }>(
      "SELECT COUNT(*) as cnt FROM notification_logs WHERE date(scheduled_for) = ? AND kind NOT IN ('global_limit')",
      day
    );
    return row?.cnt ?? 0;
  }

  private async todayRuleCount(db: SQLite.SQLiteDatabase, kind: SmartNotificationKind, date: Date): Promise<number> {
    const day = date.toISOString().slice(0, 10);
    const row = await db.getFirstAsync<{ cnt: number }>(
      'SELECT COUNT(*) as cnt FROM notification_logs WHERE kind = ? AND date(scheduled_for) = ?',
      kind, day
    );
    return row?.cnt ?? 0;
  }

  private isQuietHours(hour: number, quietStart: number, quietEnd: number): boolean {
    if (quietStart > quietEnd) {
      return hour >= quietStart || hour < quietEnd;
    }
    return hour >= quietStart && hour < quietEnd;
  }

  private buildScheduleDate(date: Date, hour: number, minute: number): Date {
    const d = new Date(date);
    d.setHours(hour || 0, minute || 0, 0, 0);
    return d;
  }

  private makeLog(
    notificationId: string,
    kind: SmartNotificationKind,
    title: string,
    body: string,
    scheduledFor: string,
    now: string,
    relatedEventId?: string,
    relatedSuggestionId?: string
  ): NotificationLog {
    return {
      id: uuidv4(),
      notificationId,
      kind,
      title,
      body,
      scheduledFor,
      relatedEventId,
      relatedSuggestionId,
      createdAt: now,
    };
  }

  private addQuietLog(date: Date, kind: string, logs: NotificationLog[], now: string): void {
    logs.push({
      id: uuidv4(),
      notificationId: 'limit_reached',
      kind: SmartNotificationKind.freeDay,
      title: 'Limit reached',
      body: `Daily notification limit reached for ${date.toISOString().slice(0, 10)}`,
      scheduledFor: now,
      createdAt: now,
    });
  }
}
